src/lib/middleware/on-error.ts

```diff
@@ -0,0 +1,55 @@
+import { Middleware, AnyContext } from '../context'
+import { CallType } from '../call-types'
+
+type ErrorHandler = (error: Error, context: AnyContext) => any
+
+/**
+ * Patches emit function to intercept errors with a handler to be able to consume or map dispatched errors on stream, before existing listeners are invoked.
+ * Error events are passed on the handler and metadata are passed on re-thrown errors.
+ * @param emitter gRPC call
+ * @param handler Error handler that can either throw an error, that will be passed on or not throw, in which case cascade is stopped and call does not end with error
+ */
+const mapError = (emitter: any, handler: ErrorHandler) => {
+  const originalEmit = emitter.emit
+  const addMeta = (e: any) => {
+    // Trailing metadata on error calls are taken from `error.metadata`, we pass on the existing trailing metadata
+    // TODO: At this point existing metadata on error is lost
+    e.metadata = emitter.trailingMetadata
+    return e
+  }
+  emitter.emit = (...args: any[]) => {
+    if (args[0] !== 'error') {
+      return originalEmit.apply(emitter, args)
+    }
+    try {
+      handler(args[1], emitter)?.catch((e: Error) =>
```

Comment:
You should be able to `await` the handler here. It works for promises and non-promises as well

Reply:
Good point, i had some issues with linter on previous implementation and then forgot about it, thx. Fixed.